### PR TITLE
chore: Prepare DAL for PHPStan update 2026-08

### PR DESCRIPTION
**Pull request by mitelg**
### 1. Why is this change necessary?

This is a smaller changeset extracted from the PHPStan update PR https://github.com/shopware/shopware/pull/17661

### 2. What does this change do, exactly?

Fixes issues in the DAL

### 4. Please link to the relevant issues (if any).
- relates #17658 

### 5. Checklist

- [ ] I have written tests and verified that they fail without my change
- [ ] I have updated developer-facing release notes if this change is **relevant** for external developers:
  - Add a short entry to `RELEASE_INFO-6.<major>.md` under “Upcoming” for informational changes, including the consequences of the change and how it affects external developers.
  - Add an `UPGRADE` section in `UPGRADE-6.<next-major>.md` for breaking changes (what/why/impact/how to adapt).
  - See the [Documenting a Release Process](https://github.com/shopware/shopware/blob/trunk/delivery-process/documenting-a-release.md) for details.
- [ ] I have written or adjusted the documentation and agent skills according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfilled them

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved validation for search filters, sorting, limits, and post-filters.
  * Added clearer error reporting for invalid filter structures, fields, values, and operators.
  * Relative-time filters now support equality operators and handle operator casing consistently.
  * Invalid relative-time intervals now return meaningful parser errors.
  * Entity existence checks handle defaults and invalid options more reliably.

* **Tests**
  * Expanded coverage for search parsing, nested filters, edge cases, and entity validation.

### CODE REVIEW

**Review by shyim — APPROVED** [review]

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php [hunk 1]
```diff
@@ -734,15 +734,72 @@ public static function providerInvalidPaging(): \Generator
...
+        try {
+            $this->requestCriteriaBuilder->fromArray($payload, new Criteria(), $this->staticDefinitionRegistry->get(ProductDefinition::class), Context::createDefaultContext());
+        } catch (SearchRequestException $e) {
+            $error = $e->getErrors()->current();
+
+            static::assertSame('FRAMEWORK__INVALID_FILTER_QUERY', $error['code']);
+            static::assertSame('The key for filter at position "1" must not be blank.', $error['detail']);
+            static::assertSame('400', $error['status']);
+            static::assertSame('/filter/1', $error['source']['pointer']);
+
```
Comment (nfortier-shopware) [inline tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:752]: what is the value of asserting it here the details of the error when we assert the exact same thing on expectExceptionObject level? 🤔

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php [hunk 2]
```diff
@@ -734,15 +734,72 @@ public static function providerInvalidPaging(): \Generator
...
+        try {
+            $this->requestCriteriaBuilder->fromArray($payload, new Criteria(), $this->staticDefinitionRegistry->get(ProductDefinition::class), Context::createDefaultContext());
+        } catch (SearchRequestException $e) {
+            $error = $e->getErrors()->current();
+
+            static::assertSame('FRAMEWORK__INVALID_FILTER_QUERY', $error['code']);
+            static::assertSame('The value for filter "name" must not be blank.', $error['detail']);
+            static::assertSame('400', $error['status']);
+            static::assertSame('/filter/name', $error['source']['pointer']);
+
```
Comment (nfortier-shopware) [inline tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:776]: Same here, expectExceptionObject is covering this already

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php [hunk 3]
```diff
@@ -752,6 +809,7 @@ public function testSimpleFilterAddsExceptionWithArrayInValue(): void
             static::assertSame('FRAMEWORK__INVALID_FILTER_QUERY', $error['code']);
             static::assertSame('The value for filter "name" must be scalar.', $error['detail']);
             static::assertSame('400', $error['status']);
+            static::assertSame('/filter/name', $error['source']['pointer']);
```
Comment (nfortier-shopware) [inline tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php]: Aaah now I understand where agent can be lost: this was necessary when the expectException( was used instead

Please remove these assertion now that expectExceptionObject is in place 👍

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php [hunk 4]
```diff
@@ -775,6 +835,108 @@ public function testFilterElementIsInvalid(): void
             static::assertSame('FRAMEWORK__INVALID_FILTER_QUERY', $error['code']);
             static::assertSame('The filter parameter has to be an array.', $error['detail']);
             static::assertSame('400', $error['status']);
+            static::assertSame('/filter/0', $error['source']['pointer']);
+
```
Comment (nfortier-shopware) [inline tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:823]: Same here no need, expectExceptionObject is enough, let's remove these assertions

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php [hunk 5]
```diff
@@ -775,6 +835,108 @@ public function testFilterElementIsInvalid(): void
...
+        try {
+            $this->requestCriteriaBuilder->fromArray($payload, new Criteria(), $this->staticDefinitionRegistry->get(ProductDefinition::class), Context::createDefaultContext());
+        } catch (SearchRequestException $e) {
+            $error = $e->getErrors()->current();
+
+            static::assertSame('FRAMEWORK__INVALID_FILTER_QUERY', $error['code']);
+            static::assertSame('The filter parameter has to be a list of filters.', $error['detail']);
+            static::assertSame('400', $error['status']);
+            static::assertSame('/filter', $error['source']['pointer']);
+
```
Comment (nfortier-shopware) [inline tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:844]: same here

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php [hunk 6]
```diff
@@ -775,6 +835,108 @@ public function testFilterElementIsInvalid(): void
...
+        try {
+            $this->requestCriteriaBuilder->fromArray($payload, new Criteria(), $this->staticDefinitionRegistry->get(ProductDefinition::class), Context::createDefaultContext());
+        } catch (SearchRequestException $e) {
+            $error = $e->getErrors()->current();
+
+            static::assertSame('FRAMEWORK__INVALID_FILTER_QUERY', $error['code']);
+            static::assertSame('The value for post-filter "name" must be scalar.', $error['detail']);
+            static::assertSame('400', $error['status']);
+            static::assertSame('/post-filter/name', $error['source']['pointer']);
+
```
Comment (nfortier-shopware) [inline tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:868]: same here

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php [hunk 7]
```diff
@@ -775,6 +835,108 @@ public function testFilterElementIsInvalid(): void
...
+        try {
+            $this->requestCriteriaBuilder->fromArray($payload, new Criteria(), $this->staticDefinitionRegistry->get(ProductDefinition::class), Context::createDefaultContext());
+        } catch (SearchRequestException $e) {
+            $error = $e->getErrors()->current();
+
+            static::assertSame('FRAMEWORK__INVALID_FILTER_QUERY', $error['code']);
+            static::assertSame('The post-filter parameter has to be an array.', $error['detail']);
+            static::assertSame('400', $error['status']);
+            static::assertSame('/post-filter/0', $error['source']['pointer']);
+
```
Comment (nfortier-shopware) [inline tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:891]: same here

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php [hunk 8]
```diff
@@ -775,6 +835,108 @@ public function testFilterElementIsInvalid(): void
...
+        try {
+            $this->requestCriteriaBuilder->fromArray($payload, new Criteria(), $this->staticDefinitionRegistry->get(ProductDefinition::class), Context::createDefaultContext());
+        } catch (SearchRequestException $e) {
+            $error = $e->getErrors()->current();
+
+            static::assertSame('FRAMEWORK__INVALID_FILTER_QUERY', $error['code']);
+            static::assertSame('The post-filter parameter has to be a list of filters.', $error['detail']);
+            static::assertSame('400', $error['status']);
+            static::assertSame('/post-filter', $error['source']['pointer']);
```
Comment (nfortier-shopware) [inline tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php]: same here

**Review by nfortier-shopware — CHANGES_REQUESTED** [review]
some tests can be simplified (they became redundant because of the upgrade from expectException to expectExceptionObject)

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php (hunk 1 above)
Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:752]: I thought the same before, but that is not correct :grimacing: 
`expectExceptionObject` is only checking class, message and code of parent exception, but the SearchRequestException has child exceptions, which are not considered. mind the `$error = $e->getErrors()->current();` calls

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php (hunk 2 above)
Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:776]: nope, see above

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php (hunk 3 above)
Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php]: nope, we can't do that, see above

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php (hunk 4 above)
Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:823]: nope, see above

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php (hunk 6 above)
Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:868]: nope, see above

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php (hunk 7 above)
Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:891]: nope, see above

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php (hunk 8 above)
Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php]: nope, see above

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php (hunk 5 above)
Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:844]: nope, see above

File: tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php (hunk 1 above)
Comment (nfortier-shopware) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:752]: Yes that's correct, then no way to avoid it. 
We can at least use  `DataAbstractionLayerException::invalidFilterQuery(` to form the exception instead of hardcode status etc.

Like in https://github.com/shopware/shopware/blob/b37c3350bfa23cbbced54a2dd6c1f145a02c59ab/tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php#L471-L474

Not sure we can aggregate these new tests with a dataprovider, but that would be ideal too

Comment (mitelg) [reply tests/unit/Core/Framework/DataAbstractionLayer/Search/RequestCriteriaBuilderTest.php:752]: refactored the whole class to make use of the exception factories and real exception objects instead of hardcoded values

**Review by nfortier-shopware — APPROVED** [review]
